### PR TITLE
Adds an s3 list prefixes operator

### PR DESCRIPTION
**Pull request by jarfgit**
- Adds an operator to return a list of prefixes from an S3 bucket
- Updates `list_prefixes()` unit test to assert on a nested dir with a prefix variable
- Removes duplicate calls to `list_keys()` that were in the `test_list_prefixes()` unit test (likely a copy/paste boo boo?)

There are two suggestion from [this conversation](https://github.com/apache/airflow/pull/8464) that I have not included here:
1. Combine or otherwise simplify `s3_list_keys()` and `s3_list_prefixes()` into one - this makes sense to me but I don't quite know how people tend to use these operators or if there is a valid argument for keeping them separate. 

2. Combining all the s3 operators into one file like [gcs.py](https://github.com/apache/airflow/blob/main/airflow/providers/google/cloud/operators/gcs.py) - this also makes sense to me, but it's not consistent with the other AWS operators. Might be worth opening a new issue to refactor them all if we want to go in this direction?

Issue Link: #[8448](https://github.com/apache/airflow/issues/8448) 

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Review by o-nikolas — COMMENTED** [review]
Thanks for the contribution!

I'm seeing a lot of duplicate code here. As called out previously in the issue, I'd like to see this as at least one operator. The hooks are a different story, which I think is outside the scope of your contribution here.

To me it makes sense to keep the current operator for backwards compatibility and also since it's name is nicely generic (i.e. it's `S3ListOperator`, not `S3ListKeysOperator`). I propose we add a new optional boolean parameter to the existing `S3ListOperator`, perhaps something like `include_subfolders` or `include_common_prefixes`. This should be false by default to maintain the current behaviour, but when true it includes common prefixes in the result payload.

The only issue with this that I see is that if folks want *only* the common prefixes, the new parameter will have to be more complex than just a boolean (your conditions in that case would be `keys`, `common_prefixes`, `all`). **Note**: if we consider that usecase uncommon then another option the users could do is call the operator twice; once with just keys and a second time with keys and subfolders and then do a set difference to get just the subfolders.

**Comment by jarfgit** [comment]
@o-nikolas I like what you're thinking. Thanks for giving me more context. I'll see what I can do.

**Comment by potiuk** [comment]
Hey @jarfgit  - any news :) ?

**Comment by iostreamdoth** [comment]
I propose if we can rename parameter to `recursive` in place of (`include_subfolders or `include_common_prefixes`) in `s3listoperator` to be in concordance with `aws s3 ls` command line tool options. Default recursive is true to retain the current behavior of operator.
thought @o-nikolas @jarfgit

**Comment by jarfgit** [comment]
@potiuk still working on it. I'm actually on the airflow team at astronomer and got pulled into something else. I'll circle back on this shortly. Sorry for the delay!

**Comment by o-nikolas** [comment]
> I propose if we can rename parameter to `recursive` in place of (`include_subfolders` or `include_common_prefixes`) in `s3listoperator` to be in concordance with `aws s3 ls` command line tool options. Default recursive is true to retain the current behavior of operator.
> thoughts @o-nikolas @jarfgit ?

I think that's a perfectly fine name, but I don't like the reasoning of tying it to the `aws s3 ls` command. If anything I think we should be tying to choose naming that closely follows the boto api that this operator wraps (`list_objects_v2`) and that api uses the common S3 language of `prefixes`. But again, I think `recursive` is expressive enough and I'm happy to agree to that :)

**Comment by jarfgit** [comment]
@o-nikolas @iostreamdoth @potiuk Ok, at long last I've updated this pull request:

* Refactored `S3ListOperator` to take a `recursive` parameter
* Refactored the S3 list keys hook to return both a list of keys and list of common prefixes (I know this is out of scope for the issue, but making unnecessary API calls bothers me and it was easy enough to do)
* If `recursive == True` then we add the prefixes to the list of keys returned by the operator
* Ensured `recursive` defaults to `False` and _shouldn't_ present a breaking change
* Added unit tests to both the `s3_list_keys` and `s3_list_prefixes` hooks to make sure I (and presumably future contributors) fully understood what the `delimiter` and `prefix` args were doing and how various combinations affected what was returned when `recursive == True`


The above assumes the following:

* The user wants to retrieve both keys and prefixes using the same optional params (i.e. `delimiter` and `prefix`). 
* The user can distinguish keys from prefixes from the list returned by the operator. I would assume that keys have a file extension and prefixes would include the delimiter... but it seems possible that keys may not _always_ have a file extension and I'm basing my understanding of the prefixes containing the delimiter on the unit tests.


So I have this question:

If we want to refactor the operator to return both prefixes and keys but a user might want to use different optional params between keys and prefixes, I don't see an alternative other than requiring the user to use the operator twice with different params. With this in mind, is there a valid argument to have a dedicated `S3ListPrefixes` operator after all?

**Comment by o-nikolas** [comment]
> * The user can distinguish keys from prefixes from the list returned by the operator. I would assume that keys have a file extension and prefixes would include the delimiter... but it seems possible that keys may not _always_ have a file extension and I'm basing my understanding of the prefixes containing the delimiter on the unit tests.

This doesn't sound overly confident 😆

> 
> So I have this question:
> 
> If we want to refactor the operator to return both prefixes and keys but a user might want to use different optional params between keys and prefixes, I don't see an alternative other than requiring the user to use the operator twice with different params. With this in mind, is there a valid argument to have a dedicated `S3ListPrefixes` operator after all?

It does sound like you're finding significant friction here. Two paths forward I see are:
1. Change the type of the new parameter to something more advanced that will allow the user to specify exactly what they want out of the command (something like a string or an enum, that can be `keys`, `prefixes`, or `all`). This doesn't get around the issue of having two calls for different optional params though. But, at that point, the user really should have two separate calls if they are querying for a completely distinct set of results IMHO.
1. Or, just bail on a single operator and go back to having two as you suggest.

What do others think?

File: tests/providers/amazon/aws/hooks/test_s3.py [hunk 1]
```diff
@@ -124,11 +124,14 @@ def test_list_prefixes(self, s3_bucket):
...
+        bucket.put_object(Key='dir/sub_dir/c', Body=b'c')
 
         assert [] == hook.list_prefixes(s3_bucket, prefix='non-existent/')
+        assert [] == hook.list_prefixes(s3_bucket)
         assert ['dir/'] == hook.list_prefixes(s3_bucket, delimiter='/')
-        assert ['a'] == hook.list_keys(s3_bucket, delimiter='/')
-        assert ['dir/b'] == hook.list_keys(s3_bucket, prefix='dir/')
+        assert [] == hook.list_prefixes(s3_bucket, prefix='dir/')
+        assert ['dir/sub_dir/'] == hook.list_prefixes(s3_bucket, delimiter='/', prefix='dir/')
+        assert [] == hook.list_prefixes(s3_bucket, prefix='dir/sub_dir/')
```
Comment (uranusjr) [inline tests/providers/amazon/aws/hooks/test_s3.py:134]: Are these changes related to the operator addition?

File: airflow/providers/amazon/aws/operators/s3_list_prefixes.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    def execute(self, context):
+        hook = S3Hook(aws_conn_id=self.aws_conn_id, verify=self.verify)
+
+        self.log.info(
+            'Getting the list of subfolders from bucket: %s in prefix: %s (Delimiter {%s)',
+            self.bucket,
+            self.prefix,
+            self.delimiter,
+        )
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/operators/s3_list_prefixes.py]:
```suggestion
        self.log.info(
            'Getting the list of subfolders from bucket: %s in prefix: %s (Delimiter %s)',
            self.bucket,
            self.prefix,
            self.delimiter,
        )
```

Typo? 🙂

File: tests/providers/amazon/aws/hooks/test_s3.py (hunk 1 above)
Comment (jarfgit) [reply tests/providers/amazon/aws/hooks/test_s3.py:134]: The calls to `list_keys` were in the `test_list_prefixes` unit test. I didn't immediately see why these calls were there and assumed it might have been a copy / past error. Am I missing something?

File: airflow/providers/amazon/aws/operators/s3_list_prefixes.py (hunk 2 above)
Comment (jarfgit) [reply airflow/providers/amazon/aws/operators/s3_list_prefixes.py]: Oops :)

**Comment by o-nikolas** [comment]
> @o-nikolas I went ahead and reverted back to having two distinct operators - if for no other reason than to keep this PR / convo alive. If I had a better understanding of s3 list keys / prefixes the use cases I would feel more confident exploring the first path you outlined above.
> 
> That being said, still open to workshopping this :)

Sorry I've been on vacation for the past three weeks:

Sounds good to me, I think that's perfectly valid :) I'll try have a review sometime this week.

File: airflow/providers/amazon/aws/operators/s3_list_prefixes.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    template_fields: Iterable[str] = ('bucket', 'prefix', 'delimiter')
+    ui_color = '#ffd700'
+
+    def __init__(
+        self,
+        *,
+        bucket: str,
+        prefix: str = '',
+        delimiter: str = '',
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/operators/s3_list_prefixes.py]: Does it really make sense to allow the user to not provide these values?

File: tests/providers/amazon/aws/operators/test_s3_list_prefixes.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        mock_hook.return_value.list_prefixes.return_value = MOCK_SUBFOLDERS
+
+        operator = S3ListPrefixesOperator(task_id=TASK_ID, bucket=BUCKET, prefix=PREFIX, delimiter=DELIMITER)
+
+        subfolders = operator.execute(None)
+
+        mock_hook.return_value.list_prefixes.assert_called_once_with(
+            bucket_name=BUCKET, prefix=PREFIX, delimiter=DELIMITER
+        )
+        assert sorted(subfolders) == sorted(MOCK_SUBFOLDERS)
```
Comment (o-nikolas) [inline tests/providers/amazon/aws/operators/test_s3_list_prefixes.py]: Do we need the sort if there is only one item in `MOCK_SUBFOLDERS`?

File: airflow/providers/amazon/aws/operators/s3_list_prefixes.py (hunk 3 above)
Comment (jarfgit) [reply airflow/providers/amazon/aws/operators/s3_list_prefixes.py]: It doesn't :)

I'll admit this was a lazy copy / paste from S3List, so I updated the `prefix` and `delimiter` there as well.

File: tests/providers/amazon/aws/operators/test_s3_list_prefixes.py (hunk 4 above)
Comment (jarfgit) [reply tests/providers/amazon/aws/operators/test_s3_list_prefixes.py]: We do not, updated in the latest push.

File: airflow/providers/amazon/aws/operators/s3_list_prefixes.py (hunk 3 above)
Comment (jarfgit) [reply airflow/providers/amazon/aws/operators/s3_list_prefixes.py]: @o-nikolas github is not letting me re-request a review from you for some reason, so pinging here.

**Review by uranusjr — APPROVED** [review]

File: airflow/providers/amazon/aws/operators/s3_list.py [hunk 5]
```diff
@@ -72,8 +72,8 @@ def __init__(
         self,
         *,
         bucket: str,
-        prefix: str = '',
-        delimiter: str = '',
+        prefix: str,
+        delimiter: str,
```
Comment (kaxil) [inline airflow/providers/amazon/aws/operators/s3_list.py]: Is this change required?

Comment (jarfgit) [reply airflow/providers/amazon/aws/operators/s3_list.py]: Actually no. I copied this code for the S3ListPrefixes operator - where you'd presumably want these to be explicit values, but that doesn't necessarily apply to S3List. Let me revert this and push it up. Good catch :)

Comment (kaxil) [reply airflow/providers/amazon/aws/operators/s3_list.py]:
```suggestion
        prefix: str = '',
        delimiter: str = '',
```

**Review by kaxil — APPROVED** [review]

File: airflow/providers/amazon/aws/operators/s3_list_prefixes.py (hunk 3 above)
Comment (o-nikolas) [reply airflow/providers/amazon/aws/operators/s3_list_prefixes.py]: Hey, sorry I missed this message. Code looks good, and I see that it's merged. Congrats :)
